{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dissected and reproduced the baseline script\n",
    "----\n",
    "\n",
    "\n",
    "I cut the originial loader class (which takes around 6-8 hours for me to run) into separate loader functions for the different columns of the input.\n",
    "\n",
    "I have done some small and easy parallelization (3-10x faster).\n",
    "\n",
    "Some parts are not complete, and has to be understood in the baseline.\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os, sys,time\n",
    "from multiprocessing import Pool\n",
    "from multiprocessing import sharedctypes\n",
    "from numpy import ctypeslib\n",
    "from functools import partial\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from scipy.stats.mstats import mquantiles\n",
    "\n",
    "import pysam\n",
    "import pyBigWig\n",
    "from Bio import SeqIO\n",
    "\n",
    "from pyDNAbinding.binding_model import DNASequence, PWMBindingModel, DNABindingModels, load_binding_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#paths\n",
    "TRAIN_TSV_BASE_DIR = \"/mnt/vdisk/data/synapse/encodeChallenge_Data/422/10035422/ChIPseq/labels/\"\n",
    "DNASE_IDR_PEAKS_BASE_DIR = \"/mnt/vdisk/data/synapse/encodeChallenge_Data/251/10253251/essential_training_data/DNASE/peaks/conservative/\"\n",
    "DNASE_FOLD_COV_DIR = \"/mnt/vdisk/data/synapse/encodeChallenge_Data/251/10253251/essential_training_data/DNASE/fold_coverage_wiggles/\"\n",
    "REF_GENOME=\"hg19.genome.fa\"\n",
    "\n",
    "#get all cell line names (imr90?)\n",
    "ALL_CELL_LINES=[x.split('.')[1] for x in os.listdir(DNASE_FOLD_COV_DIR)]\n",
    "\n",
    "#do not load all for testing\n",
    "MAX_N_ROWS=int(1e5)\n",
    "\n",
    "#use this factor now\n",
    "t_factor='CTCF'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def my_timer(f,*args):\n",
    "    \"\"\"Estimate execution time for whole genome.\"\"\"\n",
    "    start=time.time()\n",
    "    f(*args)\n",
    "    print 'Estimated time for whole genome:',\n",
    "    print int((time.time()-start)*56e6/MAX_N_ROWS),'s'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The label data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated time for whole genome: 182 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A549</th>\n",
       "      <th>H1-hESC</th>\n",
       "      <th>HeLa-S3</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>IMR-90</th>\n",
       "      <th>K562</th>\n",
       "      <th>MCF-7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr10</th>\n",
       "      <th>5000350</th>\n",
       "      <th>5000550</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000400</th>\n",
       "      <th>5000600</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000450</th>\n",
       "      <th>5000650</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000500</th>\n",
       "      <th>5000700</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000550</th>\n",
       "      <th>5000750</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      A549 H1-hESC HeLa-S3 HepG2 IMR-90 K562 MCF-7\n",
       "chr   start   stop                                                \n",
       "chr10 5000350 5000550    0       0       0     0      0    0     0\n",
       "      5000400 5000600    0       0       0     0      0    0     0\n",
       "      5000450 5000650    0       0       0     0      0    0     0\n",
       "      5000500 5000700    0       0       0     0      0    0     0\n",
       "      5000550 5000750    0       0       0     0      0    0     0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_labels(t_factor,\n",
    "                    nrows=MAX_N_ROWS,\n",
    "                    label_dir=TRAIN_TSV_BASE_DIR):\n",
    "    \"\"\"Load the labels for a transcription factor for all cell lines.\"\"\"\n",
    "    labels_fname = label_dir + t_factor + \".train.labels.tsv.gz\"\n",
    "    #load the original\n",
    "    temp_df=pd.read_table(labels_fname,header=0,index_col=(0,1,2),nrows=nrows)\n",
    "    #change U,A,B to U,A->0 B->1!!\n",
    "    temp_df[(temp_df=='U') | (temp_df=='A')]=0\n",
    "    temp_df[temp_df=='B']=1\n",
    "    return temp_df\n",
    "    \n",
    "\n",
    "my_timer(load_labels,t_factor)\n",
    "\n",
    "load_labels(t_factor).tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the positions which have to predicted\n",
    "- could be get from the above, but if needed separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated time for whole genome: 113 s\n"
     ]
    }
   ],
   "source": [
    "def load_index(t_factor='CTCF',\n",
    "               nrows=MAX_N_ROWS,\n",
    "               label_dir=TRAIN_TSV_BASE_DIR):\n",
    "    \"\"\"Load the positions which has to predicted as an index.\"\"\"\n",
    "    labels_fname = label_dir + t_factor + \".train.labels.tsv.gz\"\n",
    "    return pd.read_table(labels_fname,header=0,index_col=(0,1,2),\n",
    "                         usecols=(0,1,2),nrows=nrows).index\n",
    "\n",
    "my_timer(load_index)\n",
    "\n",
    "idx=load_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNASE peak data\n",
    "- How will we use this data?\n",
    "    - it needs to be intersected with the label indices\n",
    "- todo: understand how the baseline uses it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated time for whole genome: 92 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>chr10</td>\n",
       "      <td>112064555</td>\n",
       "      <td>112065386</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>17.57322</td>\n",
       "      <td>237.07674</td>\n",
       "      <td>233.09158</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>chr2</td>\n",
       "      <td>198380260</td>\n",
       "      <td>198381050</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>17.57322</td>\n",
       "      <td>131.93927</td>\n",
       "      <td>128.62138</td>\n",
       "      <td>464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>chr7</td>\n",
       "      <td>44887597</td>\n",
       "      <td>44888359</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>17.57426</td>\n",
       "      <td>305.57327</td>\n",
       "      <td>301.09747</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>chr2</td>\n",
       "      <td>134192461</td>\n",
       "      <td>134193257</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>17.57469</td>\n",
       "      <td>176.14229</td>\n",
       "      <td>172.54501</td>\n",
       "      <td>601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>chr11</td>\n",
       "      <td>67113263</td>\n",
       "      <td>67114300</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>.</td>\n",
       "      <td>17.57478</td>\n",
       "      <td>325.62643</td>\n",
       "      <td>320.99326</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0          1          2  3     4  5         6          7  \\\n",
       "99995  chr10  112064555  112065386  .  1000  .  17.57322  237.07674   \n",
       "99996   chr2  198380260  198381050  .  1000  .  17.57322  131.93927   \n",
       "99997   chr7   44887597   44888359  .  1000  .  17.57426  305.57327   \n",
       "99998   chr2  134192461  134193257  .  1000  .  17.57469  176.14229   \n",
       "99999  chr11   67113263   67114300  .  1000  .  17.57478  325.62643   \n",
       "\n",
       "               8    9  \n",
       "99995  233.09158  273  \n",
       "99996  128.62138  464  \n",
       "99997  301.09747  291  \n",
       "99998  172.54501  601  \n",
       "99999  320.99326  381  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_dnase_peaks(cell_line,\n",
    "                     nrows=MAX_N_ROWS,\n",
    "                     dnase_peak_dir=DNASE_IDR_PEAKS_BASE_DIR):\n",
    "    \"\"\"Load dnase peak data for a transcription factor and a cell line.\"\"\"\n",
    "    dnase_peak_fname = dnase_peak_dir + 'DNASE.' + cell_line + '.conservative.narrowPeak.gz'\n",
    "    return pd.read_table(dnase_peak_fname,header=None,nrows=nrows)\n",
    "\n",
    "\n",
    "my_timer(load_dnase_peaks,'A549')\n",
    "\n",
    "load_dnase_peaks('A549').tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The DNASE raw fold coverage data\n",
    "\n",
    "- parellization made it 3x faster\n",
    "    - bigwigs cannot be serialized, so i just create and close them in every process which creates an overhead.\n",
    "    - probably could be faster, but there are slower stuff so it doesnt matter that much\n",
    "    \n",
    "- if we want to load it for many cell lines it is slower"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated time for whole genome: 1320 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A549_dnase_fc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr10</th>\n",
       "      <th>5000350</th>\n",
       "      <th>5000550</th>\n",
       "      <td>0.227958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000400</th>\n",
       "      <th>5000600</th>\n",
       "      <td>0.257754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000450</th>\n",
       "      <th>5000650</th>\n",
       "      <td>0.240680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000500</th>\n",
       "      <th>5000700</th>\n",
       "      <td>0.256136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000550</th>\n",
       "      <th>5000750</th>\n",
       "      <td>0.261165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       A549_dnase_fc\n",
       "chr   start   stop                  \n",
       "chr10 5000350 5000550       0.227958\n",
       "      5000400 5000600       0.257754\n",
       "      5000450 5000650       0.240680\n",
       "      5000500 5000700       0.256136\n",
       "      5000550 5000750       0.261165"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#global shared result array... is there a better way?\n",
    "fold_cov = sharedctypes.RawArray('d', len(idx))\n",
    "\n",
    "def load_dnase_fold_cov(cell_line,\n",
    "                        n_proc=12,\n",
    "                        dnase_fold_cov_dir=DNASE_FOLD_COV_DIR ):\n",
    "    \"\"\"Load the raw fold coverage for a t factor and cell line.\"\"\"\n",
    "    dnase_fold_cov_fname = dnase_fold_cov_dir + 'DNASE.'+cell_line+'.fc.signal.bigwig'\n",
    "    \n",
    "    #partially apply the scorer function\n",
    "    part_get_dnase_fold_cov_from_region=partial(get_dnase_fold_cov_from_region,\n",
    "                                                fn=dnase_fold_cov_fname)\n",
    "    #parallel execute it\n",
    "    Pool(n_proc).map(part_get_dnase_fold_cov_from_region,xrange(len(idx)))\n",
    "\n",
    "    #return a dataframe\n",
    "    return pd.DataFrame({cell_line+'_dnase_fc' : fold_cov},index=idx)\n",
    "\n",
    "def get_dnase_fold_cov_from_region(i,fn=None):\n",
    "    \"\"\"Get raw dnase fold coverage for a region.\"\"\"\n",
    "    contig,start,stop=idx[i]\n",
    "    bigwig_f=pyBigWig.open(fn)\n",
    "    fold_cov[i]=bigwig_f.stats(contig,start,stop)[0]\n",
    "    bigwig_f.close()\n",
    "    return\n",
    "    \n",
    "\n",
    "my_timer(load_dnase_fold_cov,'A549')\n",
    "load_dnase_fold_cov('A549').tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the motif scores\n",
    "\n",
    "- genome need to be read only once its like 2 minutes, i do it before once\n",
    "\n",
    "- parallelization made ~10x faster, still long time\n",
    "    - genome gets copied into every process but still faster than pysam which is disk reading based"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56 s\n"
     ]
    }
   ],
   "source": [
    "start=time.time()\n",
    "with open(REF_GENOME, \"rU\") as h:\n",
    "    ref_genome=SeqIO.to_dict(SeqIO.parse(h, \"fasta\"))\n",
    "print int(time.time()-start),'s'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Estimated time for whole genome: 4095 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>CTCF_motif_mean</th>\n",
       "      <th>CTCF_motif_max</th>\n",
       "      <th>CTCF_motif_q99</th>\n",
       "      <th>CTCF_motif_q95</th>\n",
       "      <th>CTCF_motif_q90</th>\n",
       "      <th>CTCF_motif_q75</th>\n",
       "      <th>CTCF_motif_q50</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr10</th>\n",
       "      <th>50406100</th>\n",
       "      <th>50406300</th>\n",
       "      <td>6.160816</td>\n",
       "      <td>7.488257</td>\n",
       "      <td>7.340118</td>\n",
       "      <td>7.066297</td>\n",
       "      <td>6.943441</td>\n",
       "      <td>6.771597</td>\n",
       "      <td>6.504157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50406150</th>\n",
       "      <th>50406350</th>\n",
       "      <td>6.227698</td>\n",
       "      <td>7.340257</td>\n",
       "      <td>7.168880</td>\n",
       "      <td>6.996001</td>\n",
       "      <td>6.905165</td>\n",
       "      <td>6.689737</td>\n",
       "      <td>6.437557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50406200</th>\n",
       "      <th>50406400</th>\n",
       "      <td>5.868767</td>\n",
       "      <td>7.072957</td>\n",
       "      <td>7.069085</td>\n",
       "      <td>6.932773</td>\n",
       "      <td>6.799685</td>\n",
       "      <td>6.577077</td>\n",
       "      <td>6.336956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50406250</th>\n",
       "      <th>50406450</th>\n",
       "      <td>5.795236</td>\n",
       "      <td>7.072957</td>\n",
       "      <td>7.063795</td>\n",
       "      <td>6.844825</td>\n",
       "      <td>6.695881</td>\n",
       "      <td>6.481957</td>\n",
       "      <td>6.283857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50406300</th>\n",
       "      <th>50406500</th>\n",
       "      <td>5.648257</td>\n",
       "      <td>7.072957</td>\n",
       "      <td>7.039548</td>\n",
       "      <td>6.810221</td>\n",
       "      <td>6.687405</td>\n",
       "      <td>6.525657</td>\n",
       "      <td>6.306056</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         CTCF_motif_mean  CTCF_motif_max  CTCF_motif_q99  \\\n",
       "chr   start    stop                                                        \n",
       "chr10 50406100 50406300         6.160816        7.488257        7.340118   \n",
       "      50406150 50406350         6.227698        7.340257        7.168880   \n",
       "      50406200 50406400         5.868767        7.072957        7.069085   \n",
       "      50406250 50406450         5.795236        7.072957        7.063795   \n",
       "      50406300 50406500         5.648257        7.072957        7.039548   \n",
       "\n",
       "                         CTCF_motif_q95  CTCF_motif_q90  CTCF_motif_q75  \\\n",
       "chr   start    stop                                                       \n",
       "chr10 50406100 50406300        7.066297        6.943441        6.771597   \n",
       "      50406150 50406350        6.996001        6.905165        6.689737   \n",
       "      50406200 50406400        6.932773        6.799685        6.577077   \n",
       "      50406250 50406450        6.844825        6.695881        6.481957   \n",
       "      50406300 50406500        6.810221        6.687405        6.525657   \n",
       "\n",
       "                         CTCF_motif_q50  \n",
       "chr   start    stop                      \n",
       "chr10 50406100 50406300        6.504157  \n",
       "      50406150 50406350        6.437557  \n",
       "      50406200 50406400        6.336956  \n",
       "      50406250 50406450        6.283857  \n",
       "      50406300 50406500        6.306056  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the aggregate colnames\n",
    "aggregate_region_scores_labels = [\"motif_mean\", \"motif_max\", \"motif_q99\",\n",
    "                                  \"motif_q95\", \"motif_q90\", \"motif_q75\", \"motif_q50\"]\n",
    "def aggregate_region_scores(scores,\n",
    "                            quantile_probs = [0.99, 0.95, 0.90, 0.75, 0.50]):\n",
    "    \"\"\"Return aggregate scores of all scores from a region.\"\"\"\n",
    "    rv = [scores.mean()/len(scores), scores.max()]\n",
    "    rv.extend(mquantiles(scores, prob=quantile_probs))\n",
    "    return rv\n",
    "\n",
    "#global shared result array... is there a better way?\n",
    "shape=(len(idx),len(aggregate_region_scores_labels))\n",
    "motif_scores = sharedctypes.RawArray('d', shape[0]*shape[1])\n",
    "\n",
    "def load_motif_scores(t_factor,\n",
    "                      n_proc=12):\n",
    "    \"\"\"Load aggregate motif scores for a t_factor for intervals in the index.\"\"\"\n",
    "    #load binding models\n",
    "    binding_models = load_binding_models(\"models.yaml\")\n",
    "    model = binding_models.get_from_tfname(t_factor)\n",
    "    \n",
    "    #partially apply the scorer function\n",
    "    part_get_motif_scores_from_region=partial(\n",
    "        get_motif_scores_from_region,model=model)\n",
    "    \n",
    "    #parrallel execute it\n",
    "    Pool(n_proc).map(part_get_motif_scores_from_region,xrange(len(idx)))        \n",
    "    \n",
    "    #shared array to numpy array\n",
    "    ms=ctypeslib.as_array(motif_scores).reshape(shape)\n",
    "    \n",
    "    colnames=[t_factor+'_'+x for x in aggregate_region_scores_labels]\n",
    "    return pd.DataFrame(ms,index=idx,columns=colnames)\n",
    "    \n",
    "def get_motif_scores_from_region(i,model):\n",
    "    \"\"\"Get motif scores from a region in index.\"\"\"\n",
    "    contig, start, stop = idx[i]\n",
    "    #load seq from genome\n",
    "    seq=str(ref_genome[contig][start:stop].seq).upper()\n",
    "    #pysam is a bit slower\n",
    "    #genome = pysam.FastaFile(REF_GENOME)\n",
    "    #seq=genome.fetch(contig, start, stop+1).upper()\n",
    "    \n",
    "    #make sharred array numpy array\n",
    "    ms = ctypeslib.as_array(motif_scores).reshape(shape)\n",
    "    \n",
    "    #get aggregate motif scores\n",
    "    ms[i] = aggregate_region_scores(\n",
    "        DNASequence(seq).score_binding_sites(model[0], 'MAX'))\n",
    "    \n",
    "    return\n",
    "    \n",
    "my_timer(load_motif_scores,t_factor)\n",
    "load_motif_scores(t_factor).tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Put all together in one dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>A549_dnase_fc</th>\n",
       "      <th>H1-hESC_dnase_fc</th>\n",
       "      <th>HeLa-S3_dnase_fc</th>\n",
       "      <th>HepG2_dnase_fc</th>\n",
       "      <th>IMR-90_dnase_fc</th>\n",
       "      <th>K562_dnase_fc</th>\n",
       "      <th>MCF-7_dnase_fc</th>\n",
       "      <th>CTCF_motif_mean</th>\n",
       "      <th>CTCF_motif_max</th>\n",
       "      <th>CTCF_motif_q99</th>\n",
       "      <th>...</th>\n",
       "      <th>CTCF_motif_q90</th>\n",
       "      <th>CTCF_motif_q75</th>\n",
       "      <th>CTCF_motif_q50</th>\n",
       "      <th>A549</th>\n",
       "      <th>H1-hESC</th>\n",
       "      <th>HeLa-S3</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>IMR-90</th>\n",
       "      <th>K562</th>\n",
       "      <th>MCF-7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">chr10</th>\n",
       "      <th>5000350</th>\n",
       "      <th>5000550</th>\n",
       "      <td>0.227958</td>\n",
       "      <td>0.968319</td>\n",
       "      <td>0.215982</td>\n",
       "      <td>0.854929</td>\n",
       "      <td>0.111591</td>\n",
       "      <td>0.269978</td>\n",
       "      <td>0.543555</td>\n",
       "      <td>6.129895</td>\n",
       "      <td>7.353956</td>\n",
       "      <td>7.180013</td>\n",
       "      <td>...</td>\n",
       "      <td>6.910457</td>\n",
       "      <td>6.717757</td>\n",
       "      <td>6.471757</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000400</th>\n",
       "      <th>5000600</th>\n",
       "      <td>0.257754</td>\n",
       "      <td>0.782935</td>\n",
       "      <td>0.061195</td>\n",
       "      <td>0.496759</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.194384</td>\n",
       "      <td>0.244780</td>\n",
       "      <td>6.179895</td>\n",
       "      <td>7.353956</td>\n",
       "      <td>7.278669</td>\n",
       "      <td>...</td>\n",
       "      <td>6.927805</td>\n",
       "      <td>6.745437</td>\n",
       "      <td>6.472157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000450</th>\n",
       "      <th>5000650</th>\n",
       "      <td>0.240680</td>\n",
       "      <td>0.534555</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.226781</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.104391</td>\n",
       "      <td>0.082793</td>\n",
       "      <td>6.283457</td>\n",
       "      <td>7.410456</td>\n",
       "      <td>7.343331</td>\n",
       "      <td>...</td>\n",
       "      <td>6.992093</td>\n",
       "      <td>6.760777</td>\n",
       "      <td>6.535056</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000500</th>\n",
       "      <th>5000700</th>\n",
       "      <td>0.256136</td>\n",
       "      <td>0.255579</td>\n",
       "      <td>0.041397</td>\n",
       "      <td>0.016199</td>\n",
       "      <td>0.021598</td>\n",
       "      <td>0.014399</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.240154</td>\n",
       "      <td>7.410456</td>\n",
       "      <td>7.281059</td>\n",
       "      <td>...</td>\n",
       "      <td>6.999197</td>\n",
       "      <td>6.733356</td>\n",
       "      <td>6.521957</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000550</th>\n",
       "      <th>5000750</th>\n",
       "      <td>0.261165</td>\n",
       "      <td>0.111591</td>\n",
       "      <td>0.194384</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111591</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.210166</td>\n",
       "      <td>7.410456</td>\n",
       "      <td>7.281059</td>\n",
       "      <td>...</td>\n",
       "      <td>7.004493</td>\n",
       "      <td>6.762417</td>\n",
       "      <td>6.538856</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       A549_dnase_fc  H1-hESC_dnase_fc  HeLa-S3_dnase_fc  \\\n",
       "chr   start   stop                                                         \n",
       "chr10 5000350 5000550       0.227958          0.968319          0.215982   \n",
       "      5000400 5000600       0.257754          0.782935          0.061195   \n",
       "      5000450 5000650       0.240680          0.534555          0.000000   \n",
       "      5000500 5000700       0.256136          0.255579          0.041397   \n",
       "      5000550 5000750       0.261165          0.111591          0.194384   \n",
       "\n",
       "                       HepG2_dnase_fc  IMR-90_dnase_fc  K562_dnase_fc  \\\n",
       "chr   start   stop                                                      \n",
       "chr10 5000350 5000550        0.854929         0.111591       0.269978   \n",
       "      5000400 5000600        0.496759         0.000000       0.194384   \n",
       "      5000450 5000650        0.226781         0.000000       0.104391   \n",
       "      5000500 5000700        0.016199         0.021598       0.014399   \n",
       "      5000550 5000750        0.000000         0.111591       0.000000   \n",
       "\n",
       "                       MCF-7_dnase_fc  CTCF_motif_mean  CTCF_motif_max  \\\n",
       "chr   start   stop                                                       \n",
       "chr10 5000350 5000550        0.543555         6.129895        7.353956   \n",
       "      5000400 5000600        0.244780         6.179895        7.353956   \n",
       "      5000450 5000650        0.082793         6.283457        7.410456   \n",
       "      5000500 5000700        0.000000         6.240154        7.410456   \n",
       "      5000550 5000750        0.000000         6.210166        7.410456   \n",
       "\n",
       "                       CTCF_motif_q99  ...   CTCF_motif_q90  CTCF_motif_q75  \\\n",
       "chr   start   stop                     ...                                    \n",
       "chr10 5000350 5000550        7.180013  ...         6.910457        6.717757   \n",
       "      5000400 5000600        7.278669  ...         6.927805        6.745437   \n",
       "      5000450 5000650        7.343331  ...         6.992093        6.760777   \n",
       "      5000500 5000700        7.281059  ...         6.999197        6.733356   \n",
       "      5000550 5000750        7.281059  ...         7.004493        6.762417   \n",
       "\n",
       "                       CTCF_motif_q50  A549 H1-hESC HeLa-S3 HepG2 IMR-90 K562  \\\n",
       "chr   start   stop                                                              \n",
       "chr10 5000350 5000550        6.471757     0       0       0     0      0    0   \n",
       "      5000400 5000600        6.472157     0       0       0     0      0    0   \n",
       "      5000450 5000650        6.535056     0       0       0     0      0    0   \n",
       "      5000500 5000700        6.521957     0       0       0     0      0    0   \n",
       "      5000550 5000750        6.538856     0       0       0     0      0    0   \n",
       "\n",
       "                      MCF-7  \n",
       "chr   start   stop           \n",
       "chr10 5000350 5000550     0  \n",
       "      5000400 5000600     0  \n",
       "      5000450 5000650     0  \n",
       "      5000500 5000700     0  \n",
       "      5000550 5000750     0  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load the labels\n",
    "labels_df=load_labels(t_factor)\n",
    "\n",
    "#get the index of positions\n",
    "idx=labels_df.index\n",
    "\n",
    "#load fold coverage for all cell lines which have labels\n",
    "dnase_fc_df=[ load_dnase_fold_cov(cell_line) for cell_line in labels_df.columns]\n",
    "\n",
    "#load motif scores for the tf\n",
    "motif_score_df=load_motif_scores(t_factor)\n",
    "\n",
    "#concat tables\n",
    "small_df=pd.concat(dnase_fc_df+[motif_score_df,labels_df],axis=1)\n",
    "\n",
    "small_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----\n",
    "\n",
    "## Small training\n",
    "\n",
    "- with training on only 1 cell line and using another  cell line for blind test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#lables\n",
    "y_train=small_df['A549'].values.astype(int)\n",
    "y_test=small_df['K562'].values.astype(int)\n",
    "\n",
    "#dnase levels\n",
    "dnase_train=small_df['A549_dnase_fc']\n",
    "dnase_test=small_df['K562_dnase_fc']\n",
    "\n",
    "#drop these\n",
    "small_df.drop(['A549','K562','A549_dnase_fc','K562_dnase_fc'],\n",
    "              inplace=True,axis=1)\n",
    "\n",
    "#add the corrsponding dnase levels to all inputs\n",
    "df_train=pd.DataFrame(small_df)\n",
    "df_train['fold_cov']=dnase_train\n",
    "df_test=pd.DataFrame(small_df)\n",
    "df_test['fold_cov']=dnase_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train AUC: 1.0\n",
      "Test AUC: 0.956594919461\n"
     ]
    }
   ],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "\n",
    "clf=RandomForestClassifier(n_estimators=500,n_jobs=12)\n",
    "clf.fit(df_train,y_train)\n",
    "\n",
    "print 'Train AUC:',roc_auc_score(y_train,clf.predict_proba(df_train)[:,1])\n",
    "print 'Test AUC:',roc_auc_score(y_test,clf.predict_proba(df_test)[:,1])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
